sql/open_address_hash.h

```diff
@@ -0,0 +1,334 @@
+#pragma once
+
+#include <string.h>
+#include <my_dbug.h>
+
+
+
+template <typename key_trait, typename value_trait> class open_address_hash
+{
+public:
+  using T= typename value_trait::elem_type;
+  using find_type= typename value_trait::elem_type;
+  using erase_type= typename value_trait::elem_type;
+  using hash_value_type= typename key_trait::hash_value_type;
+  using key_type= typename key_trait::key_type;
+
+  //key_type *get_key(const T &elem) { return key_trait::get_key(elem); }
+  bool is_empty(const T &el) { return value_trait::is_empty(el); }
+  void set_null(T &el) { value_trait::set_null(el); }
+
+  open_address_hash()
+  {
+    first.set_mark(true);
+    first.set_ptr(nullptr);
+    second= nullptr;
+  }     
+
+  ~open_address_hash()
+  {
+    if (!first.mark())
+    {
+      DBUG_ASSERT(hash_array);
+      free(hash_array);
+    }
+  }
+
+private:
+  hash_value_type to_index(const hash_value_type &hash_value)
+  {
+    return hash_value & (capacity - 1);
+  }
+
+  hash_value_type hash_from_value(const T &value)
+  {
+    return key_trait::get_hash_value(value_trait::get_key(value));
+  }
+
+  bool insert_helper(const T &value)
+  {
+
+    auto hash_val= to_index(hash_from_value(value));
+
+    while (hash_array[hash_val] != nullptr)
+    {
+      if (hash_array[hash_val] == value)
+        return false;
+      hash_val= to_index(hash_val + 1);
+    }
+
+    hash_array[hash_val]= value;
+    return true;
+  };
+
+  uint rehash_subsequence(uint i)
+  {
+    for (uint j= to_index(i + 1); hash_array[j] != nullptr; j= to_index(j + 1))
+    {
+      auto temp_el= hash_array[j];
+      hash_array[j]= nullptr;
+      insert_helper(temp_el);
+      /*auto key= to_index(key_trait::get_hash_value(trait::get_key(hash_array[j])));
+      if (key <= i || key > j)
+      {
+        hash_array[i]= hash_array[j];
+        i= j;
+      }*/
+    }
+
+    return i;
+  }
+
+  bool erase_helper(const erase_type &value)
+  {
+    for (auto key= to_index(key_trait::get_hash_value(value_trait::get_key(value)));
+         hash_array[key] != nullptr; key= to_index(key + 1))
+    {
+      if (value_trait::is_equal(hash_array[key], value))
+      {
+        hash_array[key]= nullptr;
+        rehash_subsequence(key);
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  void grow(const uint _capacity)
+  {
+    uint past_capacity= capacity;
+    capacity= _capacity;
+    hash_array= (T *) realloc(hash_array, capacity * sizeof(T));
+
+    for (uint i= past_capacity; i < capacity; i++)
+    {
+      hash_array[i]= nullptr;
+    }
+
+    for (uint i= 0; i < capacity; i++)
+    {
+      if (hash_array[i] && i != to_index(hash_from_value(hash_array[i])))
+      {
+        auto temp_el= hash_array[i];
+        hash_array[i]= nullptr;
+        insert_helper(temp_el);
+      }
+    }
+  }
+
+  void shrink(const uint _capacity)
+  {
+    uint past_capacity= capacity;
+    capacity= _capacity;
+
+    for (uint i= capacity; i < past_capacity; i++)
```

Comment:
good. indeed we don't need to run it for `i < capacity` because of the number theory:)
*If* `i = k (mod 2^n) such that k < 2^{n-1}`
*Then* `i = k (mod 2^{n-1})`

and actually, if it's vice-versa, i.e. `k >= 2^{n-1}`,
then `i = k - 2^{n-1}  (mod 2^{n-1})`